From f40a34207a5644c70d21d41d2148658fa8147616 Mon Sep 17 00:00:00 2001
From: Armin Novak <armin.novak@thincast.com>
Date: Sat, 14 Mar 2026 08:38:12 +0100
Subject: [PATCH] [winpr,sam] fix reading SAM entries

* Fix broken length checks for hash strings
* Fix API definitions, add proper WINPR_ATTR_MALLOC and other attributes
* Fix SamLookupUserW, return the read user and domain names as WCHAR,
  the lengths in bytes
---
 winpr/include/winpr/sam.h  | 13 +++++++------
 winpr/libwinpr/utils/sam.c | 35 +++++++++++++++++++++--------------
 2 files changed, 28 insertions(+), 20 deletions(-)

diff --git a/winpr/include/winpr/sam.h b/winpr/include/winpr/sam.h
index 568848105..e73c96195 100644
--- a/winpr/include/winpr/sam.h
+++ b/winpr/include/winpr/sam.h
@@ -41,21 +41,22 @@ extern "C"
 {
 #endif
 
-	WINPR_ATTR_NODISCARD
+	WINPR_API void SamFreeEntry(WINPR_SAM* sam, WINPR_SAM_ENTRY* entry);
+
+	WINPR_ATTR_MALLOC(SamFreeEntry, 2)
 	WINPR_API WINPR_SAM_ENTRY* SamLookupUserA(WINPR_SAM* sam, LPCSTR User, UINT32 UserLength,
 	                                          LPCSTR Domain, UINT32 DomainLength);
-	WINPR_ATTR_NODISCARD
+	WINPR_ATTR_MALLOC(SamFreeEntry, 2)
 	WINPR_API WINPR_SAM_ENTRY* SamLookupUserW(WINPR_SAM* sam, LPCWSTR User, UINT32 UserLength,
 	                                          LPCWSTR Domain, UINT32 DomainLength);
 
 	WINPR_API void SamResetEntry(WINPR_SAM_ENTRY* entry);
-	WINPR_API void SamFreeEntry(WINPR_SAM* sam, WINPR_SAM_ENTRY* entry);
-
-	WINPR_ATTR_NODISCARD
-	WINPR_API WINPR_SAM* SamOpen(const char* filename, BOOL readOnly);
 
 	WINPR_API void SamClose(WINPR_SAM* sam);
 
+	WINPR_ATTR_MALLOC(SamClose, 1)
+	WINPR_API WINPR_SAM* SamOpen(const char* filename, BOOL readOnly);
+
 #ifdef __cplusplus
 }
 #endif
diff --git a/winpr/libwinpr/utils/sam.c b/winpr/libwinpr/utils/sam.c
index 76721832c..76a8327e7 100644
--- a/winpr/libwinpr/utils/sam.c
+++ b/winpr/libwinpr/utils/sam.c
@@ -246,15 +246,17 @@ static BOOL SamReadEntry(WINPR_SAM* sam, WINPR_SAM_ENTRY* entry)
 
 	if (LmHashLength == 32)
 	{
-		if (winpr_HexStringToBinBuffer(p[2], LmHashLength, entry->LmHash, sizeof(entry->LmHash)) !=
-		    32)
+		const size_t rc =
+		    winpr_HexStringToBinBuffer(p[2], LmHashLength, entry->LmHash, sizeof(entry->LmHash));
+		if (rc != 16)
 			return FALSE;
 	}
 
 	if (NtHashLength == 32)
 	{
-		if (winpr_HexStringToBinBuffer(p[3], NtHashLength, (BYTE*)entry->NtHash,
-		                               sizeof(entry->NtHash)) != 32)
+		const size_t rc = winpr_HexStringToBinBuffer(p[3], NtHashLength, (BYTE*)entry->NtHash,
+		                                             sizeof(entry->NtHash));
+		if (rc != 16)
 			return FALSE;
 	}
 
@@ -263,16 +265,8 @@ static BOOL SamReadEntry(WINPR_SAM* sam, WINPR_SAM_ENTRY* entry)
 
 void SamFreeEntry(WINPR_ATTR_UNUSED WINPR_SAM* sam, WINPR_SAM_ENTRY* entry)
 {
-	if (entry)
-	{
-		if (entry->UserLength > 0)
-			free(entry->User);
-
-		if (entry->DomainLength > 0)
-			free(entry->Domain);
-
-		free(entry);
-	}
+	SamResetEntry(entry);
+	free(entry);
 }
 
 void SamResetEntry(WINPR_SAM_ENTRY* entry)
@@ -368,6 +362,19 @@ WINPR_SAM_ENTRY* SamLookupUserW(WINPR_SAM* sam, LPCWSTR User, UINT32 UserLength,
 			goto fail;
 	}
 	entry = SamLookupUserA(sam, utfUser, (UINT32)userCharLen, utfDomain, (UINT32)domainCharLen);
+	if (entry)
+	{
+		free(entry->User);
+		free(entry->Domain);
+		entry->User = nullptr;
+		entry->Domain = nullptr;
+		if (User)
+			entry->User = (char*)_wcsdup(User);
+		entry->UserLength = UserLength;
+		if (Domain)
+			entry->Domain = (char*)_wcsdup(Domain);
+		entry->DomainLength = DomainLength;
+	}
 fail:
 	free(utfUser);
 	free(utfDomain);
-- 
2.53.0

